### PR TITLE
[new] promtail: add readline rate limit

### PR DESCRIPTION
**Pull request by liguozhong**
refer issue : https://github.com/grafana/loki/issues/2664
it is important feature,。
In our production environment, the cpu can be reduced from 3.5 core to 0.8 core, and business monitoring begins to stabilize。

image
image
image
image

### CODE REVIEW

**Comment by liguozhong** [comment]
@Whyeasy

**Comment by liguozhong** [comment]
@cyriltovena  review PR please ,this is important feature.

**Comment by Whyeasy** [comment]
Thanks @liguozhong. Will it also be possible to limit based on tags? So we only limit one service and not all services within a promtail instance?

**Comment by liguozhong** [comment]
> Thanks @liguozhong. Will it also be possible to limit based on tags? So we only limit one service and not all services within a promtail instance?

done.

**Comment by cyriltovena** [comment]
It feels like @Whyeasy and @liguozhong have different usage for this ?

@Whyeasy wanted to drop logs after a certain rate, @liguozhong you seem to use it to avoid spike of CPU but you keep the logs trading lag for CPU.

**Comment by cyriltovena** [comment]
> > Thanks @liguozhong. Will it also be possible to limit based on tags? So we only limit one service and not all services within a promtail instance?
> 
> done.

I think he wants a rate limiter stage not a global stage.

**Comment by cyriltovena** [comment]
@liguozhong can you expand a bit more on your usage please ?

**Comment by liguozhong** [comment]
> It feels like @Whyeasy and @liguozhong have different usage for this ?
> 
> @Whyeasy wanted to drop logs after a certain rate, @liguozhong you seem to use it to avoid spike of CPU but you keep the logs trading lag for CPU.

Logs are very important and do not need to be discarded. Local files can be cached for a long time. This PR code can work well in our monitoring

**Comment by liguozhong** [comment]
> @liguozhong can you expand a bit more on your usage please ?

We collect the envoy access log of the istio component in k8s to draw the golden metrics of service access traffic. 
And Deploy promtail in k8s through daemonset, and later found that the monitoring data is steep。

**Comment by liguozhong** [comment]
image
After investigation, we found that the vector agent has limited flow capability, 
so we tried to add the flow limit to promtail to solve the problem of inaccurate monitoring.

**Comment by liguozhong** [comment]
Because the default k8s promatil config file has 4 jobs, it is difficult for us to configure different current limiting rules for each job. But our business can accurately know the rate of access log generation per second. So global rate limiter is more suitable for promtail in k8s daemonset
`server:
  http_listen_port: 9080
  grpc_listen_port: 0
  log_level: debug
positions:
  filename: /run/promtail/positions.yaml
clients:
  - url: http://1.1.1.1:3100/loki/api/v1/push
scrape_configs:
  - job_name: kubernetes-pods-name
    kubernetes_sd_configs:
      - role: pod
    relabel_configs:
      - source_labels:
          - __meta_kubernetes_pod_label_name
        target_label: __service__
      - source_labels:
          - __meta_kubernetes_pod_node_name
        target_label: __host__
      - action: drop
        regex: ''
        source_labels:
          - __service__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
      - action: replace
        replacement: $1
        separator: /
        source_labels:
          - __meta_kubernetes_namespace
          - __service__
        target_label: job
      - action: replace
        source_labels:
          - __meta_kubernetes_namespace
        target_label: namespace
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_name
        target_label: pod
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_container_name
        target_label: container
      - replacement: /var/log/pods/*$1/*.log
        separator: /
        source_labels:
          - __meta_kubernetes_pod_uid
          - __meta_kubernetes_pod_container_name
        target_label: __path__
  - job_name: kubernetes-pods-app
    kubernetes_sd_configs:
      - role: pod
    relabel_configs:
      - action: drop
        regex: .+
        source_labels:
          - __meta_kubernetes_pod_label_name
      - source_labels:
          - __meta_kubernetes_pod_label_app
        target_label: __service__
      - source_labels:
          - __meta_kubernetes_pod_node_name
        target_label: __host__
      - action: drop
        regex: ''
        source_labels:
          - __service__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
      - action: replace
        replacement: $1
        separator: /
        source_labels:
          - __meta_kubernetes_namespace
          - __service__
        target_label: job
      - action: replace
        source_labels:
          - __meta_kubernetes_namespace
        target_label: namespace
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_name
        target_label: pod
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_container_name
        target_label: container
      - replacement: /var/log/pods/*$1/*.log
        separator: /
        source_labels:
          - __meta_kubernetes_pod_uid
          - __meta_kubernetes_pod_container_name
        target_label: __path__
  - job_name: kubernetes-pods-direct-controllers
    kubernetes_sd_configs:
      - role: pod
    relabel_configs:
      - action: drop
        regex: .+
        separator: ''
        source_labels:
          - __meta_kubernetes_pod_label_name
          - __meta_kubernetes_pod_label_app
      - action: drop
        regex: '[0-9a-z-.]+-[0-9a-f]{8,10}'
        source_labels:
          - __meta_kubernetes_pod_controller_name
      - source_labels:
          - __meta_kubernetes_pod_controller_name
        target_label: __service__
      - source_labels:
          - __meta_kubernetes_pod_node_name
        target_label: __host__
      - action: drop
        regex: ''
        source_labels:
          - __service__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
      - action: replace
        replacement: $1
        separator: /
        source_labels:
          - __meta_kubernetes_namespace
          - __service__
        target_label: job
      - action: replace
        source_labels:
          - __meta_kubernetes_namespace
        target_label: namespace
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_name
        target_label: pod
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_container_name
        target_label: container
      - replacement: /var/log/pods/*$1/*.log
        separator: /
        source_labels:
          - __meta_kubernetes_pod_uid
          - __meta_kubernetes_pod_container_name
        target_label: __path__
  - job_name: kubernetes-pods-indirect-controller
    kubernetes_sd_configs:
      - role: pod
    relabel_configs:
      - action: drop
        regex: .+
        separator: ''
        source_labels:
          - __meta_kubernetes_pod_label_name
          - __meta_kubernetes_pod_label_app
      - action: keep
        regex: '[0-9a-z-.]+-[0-9a-f]{8,10}'
        source_labels:
          - __meta_kubernetes_pod_controller_name
      - action: replace
        regex: '([0-9a-z-.]+)-[0-9a-f]{8,10}'
        source_labels:
          - __meta_kubernetes_pod_controller_name
        target_label: __service__
      - source_labels:
          - __meta_kubernetes_pod_node_name
        target_label: __host__
      - action: drop
        regex: ''
        source_labels:
          - __service__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
      - action: replace
        replacement: $1
        separator: /
        source_labels:
          - __meta_kubernetes_namespace
          - __service__
        target_label: job
      - action: replace
        source_labels:
          - __meta_kubernetes_namespace
        target_label: namespace
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_name
        target_label: pod
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_container_name
        target_label: container
      - replacement: /var/log/pods/*$1/*.log
        separator: /
        source_labels:
          - __meta_kubernetes_pod_uid
          - __meta_kubernetes_pod_container_name
        target_label: __path__
  - job_name: kubernetes-pods-static
    kubernetes_sd_configs:
      - role: pod
    relabel_configs:
      - action: drop
        regex: ''
        source_labels:
          - __meta_kubernetes_pod_annotation_kubernetes_io_config_mirror
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_label_component
        target_label: __service__
      - source_labels:
          - __meta_kubernetes_pod_node_name
        target_label: __host__
      - action: drop
        regex: ''
        source_labels:
          - __service__
      - action: labelmap
        regex: __meta_kubernetes_pod_label_(.+)
      - action: replace
        replacement: $1
        separator: /
        source_labels:
          - __meta_kubernetes_namespace
          - __service__
        target_label: job
      - action: replace
        source_labels:
          - __meta_kubernetes_namespace
        target_label: namespace
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_name
        target_label: pod
      - action: replace
        source_labels:
          - __meta_kubernetes_pod_container_name
        target_label: container
      - replacement: /var/log/pods/*$1/*.log
        separator: /
        source_labels:
          - __meta_kubernetes_pod_annotation_kubernetes_io_config_mirror
          - __meta_kubernetes_pod_container_name
        target_label: __path__
`

**Comment by Whyeasy** [comment]
> > It feels like @Whyeasy and @liguozhong have different usage for this ?
> > @Whyeasy wanted to drop logs after a certain rate, @liguozhong you seem to use it to avoid spike of CPU but you keep the logs trading lag for CPU.
> 
> Logs are very important and do not need to be discarded. Local files can be cached for a long time. This PR code can work well in our monitoring

@liguozhong I understand logs are important and we can indeed buffer a lot. But what if there is a service that keeps on spamming and the buffer only grows? At some point we would like to drop those logs for that particular service. That's the reason it would be nice to do it based on tags. This way we can say we drop logs for everything non-prod if it spams to much, but we buffer it for prod.

**Comment by liguozhong** [comment]
> > > It feels like @Whyeasy and @liguozhong have different usage for this ?
> > > @Whyeasy wanted to drop logs after a certain rate, @liguozhong you seem to use it to avoid spike of CPU but you keep the logs trading lag for CPU.
> > 
> > 
> > Logs are very important and do not need to be discarded. Local files can be cached for a long time. This PR code can work well in our monitoring
> 
> @liguozhong I understand logs are important and we can indeed buffer a lot. But what if there is a service that keeps on spamming and the buffer only grows? At some point we would like to drop those logs for that particular service. That's the reason it would be nice to do it based on tags. This way we can say we drop logs for everything non-prod if it spams to much, but we buffer it for prod.

Your usage is probably to add a stage pipeline operator like this, and I will commit another PR to handle your usage.

image

**Comment by liguozhong** [comment]
> That make sense to have 2 different feature set.
> 
> Still I'm not understanding what problem the global limit is solving for you.

Because of the limitation of the cpu limit of the pod in k8s.

1: When there is no rate limit, the cpu usage of the pod will exceed the cpu limit, causing the pod to enter a frozen state.

2: When with rate limit, the cpu usage of the pod will not exceed the cpu limit of the pod, so that the promtail pod will not enter the docker freeze state and work stably and continuously, so that there will be no data loss.

**Comment by cyriltovena** [comment]
Right ok so it's for throttling the CPU to avoid hitting the CPU limit.

File: clients/pkg/promtail/limit/config.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package limit
+
+import (
+	"flag"
+)
+
+type Config struct {
+	ReadlineRate        float64 `yaml:"readline_rate" json:"readline_rate"`
+	ReadlineBurst       int     `yaml:"readline_burst" json:"readline_burst"`
+	ReadlineRateEnabled bool    `yaml:"readline_rate_enabled,omitempty"  json:"readline_rate_enabled"`
```
Comment (cyriltovena) [inline clients/pkg/promtail/limit/config.go:10]: I think we should add a config to decided if we want to drop or not the log line. This makes it usable in multiple use case.

A bit like the stage one but global.

Comment (liguozhong) [reply clients/pkg/promtail/limit/config.go:10]: done,thanks .
image
image

File: clients/pkg/promtail/limit/config.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	ReadlineRateAsyn    bool    `yaml:"readline_rate_asyn,omitempty"  json:"readline_rate_asyn"`
+}
+
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.Float64Var(&cfg.ReadlineRate, prefix+"limit.readline-rate", 10000, "promtail readline Rate.")
+	f.IntVar(&cfg.ReadlineBurst, prefix+"limit.readline-burst", 10000, "promtail readline Burst.")
+	f.BoolVar(&cfg.ReadlineRateEnabled, prefix+"limit.readline-rate-enabled", true, "Set to false to disable readline rate limit.")
+	f.BoolVar(&cfg.ReadlineRateAsyn, prefix+"limit.readline-rate-asyn", false, "Set to true to drop log when rate limit.")
```
Comment (cyriltovena) [inline clients/pkg/promtail/limit/config.go:18]: ReadlineRateDrop ? What does Asyn stand for here ?

Comment (liguozhong) [reply clients/pkg/promtail/limit/config.go]: done,  "Drop" is better than "Asyn"  .

File: clients/pkg/promtail/limit/config.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	ReadlineBurst       int     `yaml:"readline_burst" json:"readline_burst"`
+	ReadlineRateEnabled bool    `yaml:"readline_rate_enabled,omitempty"  json:"readline_rate_enabled"`
+	ReadlineRateDrop    bool    `yaml:"readline_rate_drop,omitempty"  json:"readline_rate_drop"`
+}
+
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.Float64Var(&cfg.ReadlineRate, prefix+"limit.readline-rate", 10000, "promtail readline Rate.")
+	f.IntVar(&cfg.ReadlineBurst, prefix+"limit.readline-burst", 10000, "promtail readline Burst.")
+	f.BoolVar(&cfg.ReadlineRateEnabled, prefix+"limit.readline-rate-enabled", true, "Set to false to disable readline rate limit.")
+	f.BoolVar(&cfg.ReadlineRateDrop, prefix+"limit.readline-rate-drop", false, "Set to true to drop log when rate limit.")
```
Comment (cyriltovena) [inline clients/pkg/promtail/limit/config.go:18]:
```suggestion
	f.BoolVar(&cfg.ReadlineRateDrop, prefix+"limit.readline-rate-drop", true, "Set to true to drop log when rate limit.")
```

File: clients/pkg/promtail/limit/config.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+	ReadlineRate        float64 `yaml:"readline_rate" json:"readline_rate"`
+	ReadlineBurst       int     `yaml:"readline_burst" json:"readline_burst"`
+	ReadlineRateEnabled bool    `yaml:"readline_rate_enabled,omitempty"  json:"readline_rate_enabled"`
+	ReadlineRateDrop    bool    `yaml:"readline_rate_drop,omitempty"  json:"readline_rate_drop"`
+}
+
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.Float64Var(&cfg.ReadlineRate, prefix+"limit.readline-rate", 10000, "promtail readline Rate.")
+	f.IntVar(&cfg.ReadlineBurst, prefix+"limit.readline-burst", 10000, "promtail readline Burst.")
+	f.BoolVar(&cfg.ReadlineRateEnabled, prefix+"limit.readline-rate-enabled", true, "Set to false to disable readline rate limit.")
```
[the remark this answers is not in the document]
Comment (liguozhong) [reply clients/pkg/promtail/limit/config.go:17]: done

**Comment by liguozhong** [comment]
> > > It feels like @Whyeasy and @liguozhong have different usage for this ?
> > > @Whyeasy wanted to drop logs after a certain rate, @liguozhong you seem to use it to avoid spike of CPU but you keep the logs trading lag for CPU.
> > 
> > 
> > Logs are very important and do not need to be discarded. Local files can be cached for a long time. This PR code can work well in our monitoring
> 
> @liguozhong I understand logs are important and we can indeed buffer a lot. But what if there is a service that keeps on spamming and the buffer only grows? At some point we would like to drop those logs for that particular service. That's the reason it would be nice to do it based on tags. This way we can say we drop logs for everything non-prod if it spams to much, but we buffer it for prod.

this PR，https://github.com/grafana/loki/pull/5051

**Review by cyriltovena — APPROVED** [review]
LGTM

**Review by owen-d — APPROVED** [review]

**Review by dannykopping — APPROVED** [review]
Great contribution, thanks @liguozhong :tada:

**Comment by tablatronix** [comment]
FYI for anyone finding this, the config name changed

https://github.com/grafana/loki/pull/5707

This was driving me nuts

**Comment by liguozhong** [comment]
thanks

**Pull request merged**